### DISCUSSION TITLE
What's the best current approach for testing marimo notebooks?

### DISCUSSION THREAD

**Discussion by superlopuh** [marimo-team/marimo#573] [Q&A] [answered]
In a project I'm working on, the CI checks that our Jupyter notebooks outputs are consistent, by re-running the notebooks and checking the outputs vs the ones cached. This is very helpful to catch changes in behaviour when updating dependencies. Is there anything equivalent for marimo?

https://pypi.org/project/nbval/

**Comment by mscolnick** [comment, marked as answer]
We currently don't have a way to do this exactly how you do it with Jupyter. But we have some alternatives of testing your notebooks in CI.

1. You can run marimo as a script ([see docs here](https://docs.marimo.io/guides/scripts.html)). If your cells output anything to `stdout`, you could capture that output and snapshot
2. Again, running as a script, you can just make sure it runs correctly with a non-error exit code.
3. Since the notebooks are stored as py files, you can also type-check the file 

Right now to generate the outputs, it requires spinning up a browser. We plan to make some architectural changes in future that will allow use to do something like generate the outputs via the CLI.
